### DISCUSSION TITLE
How to count all trainable parameters for a nnx.Module

### DISCUSSION THREAD

**Discussion by jlperla** [google/flax#4310] [General]
I want to get a count of all of the trainable parameters for a module.  In pytorch I can use `model.parameters` and then recursively check if a gradient is required, etc.

How would I do this with a recursive function and filters in nnx?

**Comment by cgarciae** [comment]
Hey @jlperla, the easiest way is to use `nnx.state` to extract the `nnx.Param` instances:

```python
params = nnx.state(model, nnx.Param)
total_params = sum(np.prod(x.shape) for x in jax.tree.leaves(params), 0)
```

**Reply by Vilin97** [reply]
The above does not work directly for me, but this does:
```
params = nnx.state(model, nnx.Param)
total_params  = sum(np.prod(x.shape) for x in jax.tree_util.tree_leaves(params))
```
